### PR TITLE
Implement step-definitions for login feature

### PR DESCRIPTION
**Pull request by rijju-pdl**
This PR implements Login Feature Test with Step Definitions and Page Object Model.

### CODE REVIEW

File: tests/stepDefinitions/loginContext.js [hunk 1]
```diff
@@ -1,26 +1,39 @@
...
+const loginBtnLocator  = ".btn-block"
+const emailFieldLocator = "#spree_user_email"
+const passwordFieldLocator = "#spree_user_password"
+const paraFieldLocator = ".mb-3"
+const errorLocator = ".alert-danger"
 
-When('user {string} logs in with email {string} and password {string}', function (user, email, password) {
-  // Write code here that turns the phrase above into concrete actions
-  console.log(user, email, password)
+Given('the user {string} has navigated to the login page',  async function (string) {
```
Comment (prashant-gurung899) [inline tests/stepDefinitions/loginContext.js]:
```suggestion
Given('the user {string} has navigated to the login page',  async function (user) {
```

File: tests/stepDefinitions/loginContext.js [hunk 2]
```diff
@@ -1,26 +1,39 @@
...
+  await page.fill(emailFieldLocator, email);
+  await page.fill(passwordFieldLocator, password);
+  await page.click(loginBtnLocator);
 });
 
-When('user {string} tries to log in with following details:', function (string, dataTable) {
-  // Write code here that turns the phrase above into concrete actions
-  console.log("solved")
+Then('user {string} should be navigated to the admin panel dashboard', async function (string) {
+await page.waitForURL("http://localhost:3000/", { timeout: 5000 });
```
Comment (prashant-gurung899) [inline tests/stepDefinitions/loginContext.js]: Please indent your code accordingly
```suggestion
    await page.waitForURL("http://localhost:3000/", { timeout: 5000 });
```

File: tests/stepDefinitions/loginContext.js [hunk 3]
```diff
@@ -1,26 +1,39 @@
...
+});
+ 
+When('user {string} tries to log in with following details:', async function (string, credentials) {
```
Comment (prashant-gurung899) [inline tests/stepDefinitions/loginContext.js]:
```suggestion
When('user {string} tries to log in with following details:', async function (user, credentials) {
```

File: tests/stepDefinitions/loginContext.js [hunk 4]
```diff
@@ -1,26 +1,39 @@
...
+ const credentialList = credentials.hashes();
+ for(const credential of credentialList)
```
Comment (prashant-gurung899) [inline tests/stepDefinitions/loginContext.js]:
```suggestion
 const credentials = credentials.hashes();
 for(const credential of credentials)
```

File: tests/features/login.feature [hunk 5]
```diff
@@ -24,4 +24,4 @@ Feature: Login
            | abc@gmail.com   | abc@123     | Invalid email or password. |
            |                 | abc@13      | Invalid email or password. |
            |                 |             | Invalid email or password. |
-           |admin            | admin123456 | Invalid email or password. |
\ No newline at end of file
+        #    |admin            | admin123456 | Invalid email or password. |
```
Comment (prashant-gurung899) [inline tests/features/login.feature]:
```suggestion
```

File: tests/stepDefinitions/loginContext.js [hunk 6]
```diff
@@ -1,26 +1,39 @@
...
+const passwordFieldLocator = "#spree_user_password"
+const paraFieldLocator = ".mb-3"
+const errorLocator = ".alert-danger"
 
-When('user {string} logs in with email {string} and password {string}', function (user, email, password) {
-  // Write code here that turns the phrase above into concrete actions
-  console.log(user, email, password)
+Given('the user {string} has navigated to the login page',  async function (string) {
+  await page.goto("http://localhost:3000/admin/login");
+  expect(page.url()).toBe("http://localhost:3000/admin/login");
```
Comment (anon-pradip) [inline tests/stepDefinitions/loginContext.js]: we can use `await expect(page).toHaveURL("http://localhost:3000/admin/login");` for more simplicity

Comment (anon-pradip) [inline tests/stepDefinitions/loginContext.js]:
```suggestion
 await expect(page.url()).toBe("http://localhost:3000/admin/login");
```

File: tests/stepDefinitions/loginContext.js (hunk 4 above)
Comment (anon-pradip) [reply tests/stepDefinitions/loginContext.js:27]: we can go with:
```js
 for (const { email, password } of credentials.hashes()) {
      ....
    }
```
for more simplicity and less LOC.

File: tests/pageObjects/LoginPage.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
+class LoginPage {
+constructor() {
+    this.loginBtnLocator = "(//input[@name='commit'])[1]";
+    this.emailFieldLocator = "#spree_user_email"
+    this.passwordFieldLocator = "#spree_user_password"
+    this.errorLocator = ".alert-danger"
```
Comment (prashant-gurung899) [inline tests/pageObjects/LoginPage.js]: User term `selector`
```suggestion
    this.errorMessageSelector = ".alert-danger"
```

File: tests/pageObjects/LoginPage.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+}
+ async navigate() {
```
Comment (prashant-gurung899) [inline tests/pageObjects/LoginPage.js]:
```suggestion
 async navigateToAdminLoginPage() {
```

File: tests/pageObjects/LoginPage.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+ }
+
+ async loginInputFields(email, password ){
+    await page.fill(this.emailFieldLocator, email);
+    await page.fill(this.passwordFieldLocator,password);
+}
+
+async clickLoginBtn() {
+    await page.click(this.loginBtnLocator); 
+}
```
Comment (prashant-gurung899) [inline tests/pageObjects/LoginPage.js]: Combine these functions into one
```suggestion
 async login(email, password ){}
```

**Comment by prashant-gurung899** [comment]
Please, update your description.

**Comment by koebel** [comment]
Does this PR relates to the issue about E2E tests (https://github.com/spree/spree/issues/12235)? Linking to issues helps other developers who are considering to make open source contributions see that someone is already working on this task and avoid wasting time and effort by having multiple people do the same task  🙈

And if there is any dependency of this Draft PR with another open PR towards the original project (maybe with https://github.com/spree/spree/pull/12267), it's also helpful to link them, in particular if you are considering to add your tests as an open source contribution towards the original project at some point in time in the future.

**Pull request merged**